dcos-log/mesos/files/reader/read.go

```diff
@@ -103,58 +108,68 @@ func NewLineReader(client *http.Client, masterURL url.URL, agentID, frameworkID,
 			length = rm.offset
 		}
 
-		// TODO: split this function into smaller ones to improve the readability.
-		for {
-			ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
-			lines, delta, err := rm.read(ctx, offset, length, reverse)
-			if err != nil {
-				cancel()
-				return nil, err
-			}
+		err := calcOffset(offset, length, rm)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+	}
 
+	// guard against negative offset
+	if rm.offset < 0 {
+		rm.offset = 0
+	}
+
+	return rm, nil
+}
+
+func calcOffset(offset, length int, rm *ReadManager) error {
+	var foundLines int
+
+	for {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+		lines, delta, err := rm.read(ctx, offset, length, reverse)
+		if err != nil {
 			cancel()
+			return err
+		}
 
-			skip := rm.skip
+		cancel()
 
-			// make skip a positive number
-			if skip < 0 {
-				skip = rm.skip * -1
-			}
+		skip := rm.skip
 
-			// if the required number of lines found, we need to calculate an offset
-			if foundLines+len(lines) >= skip {
-				for i, skipped := 0, 0; skipped < skip; i++ {
-					if lines[i].Message != "" {
-						skipped++
-					}
-					rm.offset -= len(lines[i].Message) + 1
-				}
-				break
-			} else {
-				// if the current chunk contains less then requested lines, then add to a counter
-				// and continue search.
-				foundLines += len(lines)
-			}
+		// make skip a positive number
+		if skip < 0 {
+			skip = rm.skip * -1
+		}
 
-			length = chunkSize
-			offset -= chunkSize - delta
-			rm.offset = offset
-
-			// if the offset is 0 or negative value, the means we reached the top of the file.
-			// we can just set the offset to 0 and read the entire file
-			if offset < 0 {
-				rm.offset = 0
-				break
+		// if the required number of lines found, we need to calculate an offset
+		if foundLines+len(lines) >= skip {
+			for i, skipped := 0, 0; skipped < skip; i++ {
+				if lines[i].Message != "" {
+					skipped++
+				}
+				rm.offset -= len(lines[i].Message) + 1
 			}
+			break
+		} else {
```

Comment:
if you have a `break` above, do you need the `else` here?

Reply:
yeah, i will replace break with return, it just means stop the loop